### PR TITLE
Abhishek creating user roles

### PR DESCRIPTION
**Pull request by Abhishek-chouksey**

### CODE REVIEW

**Comment by Abhishek-chouksey** [comment]
abhishek-creatingUserRoles

**Comment by Abhishek-chouksey** [comment]
@googlebot i signed in!

File: pom.xml [hunk 1]
```diff
@@ -9,7 +9,7 @@
 
 	<groupId>com.apigee.edge.config</groupId>
 	<artifactId>apigee-config-maven-plugin</artifactId>
-	<version>1.3.6-SNAPSHOT</version>
```
Comment (ssvaidyanathan) [inline pom.xml]: @Abhishek-chouksey - pls do not change the version. The current version is 1.3.5, so the pom.xml will point to 1.3.6-SNAPSHOT. During the release, it will get pushed as 1.3.6 and this file will automatically be updated to 1.3.7-SNAPSHOT. So please revert this change

File: samples/APIandConfig/shared-pom.xml [hunk 2]
```diff
@@ -69,7 +69,7 @@
             <plugin>
                 <groupId>com.apigee.edge.config</groupId>
                 <artifactId>apigee-config-maven-plugin</artifactId>
-                <version>1.3.5</version>
+                <version>1.3.7</version>
```
Comment (ssvaidyanathan) [inline samples/APIandConfig/shared-pom.xml]: Should be 1.3.6

File: README.md [hunk 3]
```diff
@@ -62,6 +62,7 @@ mvn install -Ptest -Dapigee.config.options=create
   extensions			#v1.3.1 or later
   reports			#v1.3.2 or later
   references			#v1.3.4 or later
+  customroles  #v1.3.7 or later
```
Comment (ssvaidyanathan) [inline README.md]: Should be v1.3.6

File: samples/EdgeConfig/shared-pom.xml [hunk 4]
```diff
@@ -26,7 +26,7 @@
             <plugin>
                 <groupId>com.apigee.edge.config</groupId>
                 <artifactId>apigee-config-maven-plugin</artifactId>
-                <version>1.3.5</version>
+                <version>1.3.7</version>
```
Comment (ssvaidyanathan) [inline samples/EdgeConfig/shared-pom.xml]: Should be 1.3.6

File: samples/EdgeConfig/shared-pom.xml [hunk 5]
```diff
@@ -98,6 +98,13 @@
                             <goal>flowhooks</goal>
                         </goals>
                     </execution>
+                    <execution> 
+                        <id>create-config-customroles</id>
+                        <phase>install</phase>
+                        <goals>
+                            <goal>customroles</goal>
```
Comment (ssvaidyanathan) [inline samples/EdgeConfig/shared-pom.xml]: Can we call it `userroles` in stead of `customroles` matching the name of the resource in our Management API.

File: src/main/java/com/apigee/edge/config/mavenplugin/CacheMojo.java [hunk 6]
```diff
@@ -230,8 +230,8 @@ public void execute() throws MojoExecutionException, MojoFailureException {
      **/
     public static String createCache(ServerProfile profile, String cache)
             throws IOException {
-
-        HttpResponse response = RestUtil.createEnvConfig(profile, "caches", cache);
+        
+    	HttpResponse response = RestUtil.createEnvConfig(profile, "caches", cache);
```
Comment (ssvaidyanathan) [inline src/main/java/com/apigee/edge/config/mavenplugin/CacheMojo.java]: not sure why this file was modified. Can you remove this from the PR

File: src/main/java/com/apigee/edge/config/mavenplugin/CustomRoleMojo.java [hunk 7]
```diff
@@ -0,0 +1,352 @@
...
+ *
+ * @author Abhishek Chouksey
+ * @author Pallavi Tanpure
+ * @author Soudnya Nalawade
+ * @author Himanshu Sachdev
+ * @goal customroles
+ * @phase install
+ */
+
+public class CustomRoleMojo extends GatewayAbstractMojo {
```
Comment (ssvaidyanathan) [inline src/main/java/com/apigee/edge/config/mavenplugin/CustomRoleMojo.java]: can we use UserRoleMojo in stead of CustomRoleMojo ?
And all the references in the code to use UserRoles in stead of Custom Roles

File: src/main/java/com/apigee/edge/config/mavenplugin/CustomRoleMojo.java [hunk 8]
```diff
@@ -0,0 +1,352 @@
...
+import com.google.gson.JsonParseException;
+
+/**
+ * Goal to create Custom Roles in Apigee EDGE. scope: org
+ *
+ * @author Abhishek Chouksey
+ * @author Pallavi Tanpure
+ * @author Soudnya Nalawade
+ * @author Himanshu Sachdev
+ * @goal customroles
```
Comment (ssvaidyanathan) [inline src/main/java/com/apigee/edge/config/mavenplugin/CustomRoleMojo.java]: change goal to `userroles`

File: src/main/java/com/apigee/edge/config/mavenplugin/CustomRoleMojo.java [hunk 9]
```diff
@@ -0,0 +1,352 @@
...
+
+	/**
+	 * This method is used to update the user role Permissions.
+	 *
+	 * @param profile
+	 *
+	 *
+	 * @param customRole
+	 */
+	public static String updateCustomRole(ServerProfile profile, String customRole) throws IOException {
```
Comment (ssvaidyanathan) [inline src/main/java/com/apigee/edge/config/mavenplugin/CustomRoleMojo.java]: What happens when I create a custom role with 3 resource permissions and then update the json file to delete one resource. Now I run update ? What will it do ? Do you think it will remove the excluded permission from the role ? Was that tested ?

Comment (Abhishek-chouksey) [reply src/main/java/com/apigee/edge/config/mavenplugin/CustomRoleMojo.java]: It will update the  resource permissions to custom role and if we delete the resource permission from edge.json file then the resource permissions will not get deleted.
Running update command will only add new resource permissions to custom role. It will not remove the excluded permission from the role.
If we use the sync command then it will delete the excluded permission from the role. 
This behavior is in line with the behavior of the api(https://apidocs.apigee.com/management/apis/post/organizations/%7Borg_name%7D/userroles/%7Brole_name%7D/resourcepermissions)

File: pom.xml [hunk 10]
```diff
@@ -9,7 +9,7 @@
...
+	<version>1.3.6</version>
```
Comment (ssvaidyanathan) [inline pom.xml]: Should be 1.3.6-SNAPSHOT

File: src/main/java/com/apigee/edge/config/mavenplugin/CustomRoleMojo.java (hunk 9 above)
Comment (ssvaidyanathan) [reply src/main/java/com/apigee/edge/config/mavenplugin/CustomRoleMojo.java]: ok.. lets add that to the README

**Comment by Abhishek-chouksey** [comment]
I have changed my password and updated pom.xml. Thanks for your review, Please merge my pull request. Thank you.

File: pom.xml (hunk 10 above)
Comment (Abhishek-chouksey) [reply pom.xml]: This is taken care of now. Thanks
@ssvaidyanathan

**Comment by ssvaidyanathan** [comment]
@Abhishek-chouksey - have approved and merged the PR. Will release this plugin to Maven repo as soon as I can

Thanks again for the PR. Appreciate it

**Comment by ssvaidyanathan** [comment]
@Abhishek-chouksey - just released the Maven version v1.3.6 with your changes. I had to make changes to the UserRoleMojo as it had multiple custom role reference. Please test the plugin with this version of the plugin

Thanks once again

**Comment by Abhishek-chouksey** [comment]
@ssvaidyanathan Thanks for accepting my pull request and releasing Maven version v1.3.6. 
I have tested the plugin with this version it's working as expected.

Thanks again
